nft_ingester/src/consistency_bg_job.rs

```diff
@@ -0,0 +1,743 @@
+    grand_epoch: u16,
+    we: &impl BbgmChecksumServiceApi,
+    peer: &impl BbgmChecksumServiceApi,
+    metrics: &Peer2PeerConsistencyMetricsConfig,
+) -> Vec<BbgmChangeRecord> {
+    let mut result = Vec::new();
+    let Ok(peer_ge_chksms) = peer
+        .list_grand_epoch_checksums(grand_epoch, None, None)
+        .await
+    else {
+        return Vec::new();
+    };
+
+    let start = Instant::now();
+    let my_ge_chksms = match we.list_grand_epoch_checksums(grand_epoch, None, None).await {
+        Ok(v) => v,
+        Err(e) => {
+            tracing::error!("Error reading grand epochs from DB: {}", e.to_string());
+            return Vec::new();
+        }
+    };
+    metrics
+        .db_bubblegum_get_grand_epochs_latency
+        .observe(start.elapsed().as_secs_f64());
+
+    let ge_cmp_res = cmp(&my_ge_chksms, &peer_ge_chksms);
+    let ge_trees_to_check = ge_cmp_res
+        .we_miss
+        .iter()
+        .chain(ge_cmp_res.different.iter())
+        .map(|&a| a.tree_pubkey)
+        .collect::<Vec<_>>();
+    for tree_pk in ge_trees_to_check {
+        let Ok(peer_e_chksms) = peer.list_epoch_checksums(grand_epoch, tree_pk).await else {
+            continue;
+        };
+
+        let start = Instant::now();
+        let my_e_chksms = match we.list_epoch_checksums(grand_epoch, tree_pk).await {
+            Ok(v) => v,
+            Err(e) => {
+                tracing::error!("Error reading epochs from DB: {}", e.to_string());
+                return result;
+            }
+        };
+        metrics
+            .db_bubblegum_get_epochs_latency
+            .observe(start.elapsed().as_secs_f64());
+
+        let e_cmp_res = cmp(&my_e_chksms, &peer_e_chksms);
+        let epochs_to_check = e_cmp_res
+            .we_miss
+            .iter()
+            .chain(e_cmp_res.different.iter())
+            .map(|&a| (a.epoch, a.tree_pubkey))
+            .collect::<Vec<_>>();
+        for (epoch, tree_pubkey) in epochs_to_check {
+            let Ok(peer_changes) = peer
+                .list_epoch_changes(epoch, tree_pubkey, None, None)
+                .await
+            else {
+                continue;
+            };
+
+            let start = Instant::now();
+            let my_changes = match we.list_epoch_changes(epoch, tree_pubkey, None, None).await {
+                Ok(v) => v,
+                Err(e) => {
+                    tracing::error!("Error reading bubblegum changes from DB: {}", e.to_string());
+                    return result;
+                }
+            };
+            metrics
+                .db_bubblegum_get_changes_latency
+                .observe(start.elapsed().as_secs_f64());
+
+            let changes_cmp_res = cmp(&my_changes, &peer_changes);
+            result.extend(changes_cmp_res.we_miss.into_iter().map(|a| a.to_owned()));
+            // TODO: track different
+        }
+    }
+
+    result
+}
+
+#[allow(clippy::while_let_on_iterator)]
+async fn handle_missing_bbgm_changes(
```

Comment:
Point to discuss: a consensus mechanism 

If we have many hosts in the network we need to think about not just fetching any updates from each host but finding consensus between them and rejecting updates from some hosts if they contradict the majority of the network

Reply:
Right, this could be a good thing for interaction with non-trusted peers, but we've decided to not make it as part of the initial implementation.